From 8c07fbd1cca5cf7c56b8b7d1996a935bd5743bf4 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 20 Nov 2024 18:07:17 +0100
Subject: [PATCH] [DogeCash] Add `auxpow` to `CompressedHeader` (#56)

Currently, this quietly drops the auxpow, which is incorrect.

Note this mostly voids the idea behind CompressedHeader, but there's no simple way of fixing this.

(cherry picked from commit 16f2ffac46d1eedb185dbca4962af4bd66f0b9ad)

diff --git a/src/headerssync.cpp b/src/headerssync.cpp
index 3baf50e894fb38e861066ee3911d943e539a750e..e1552c81be22f5674c8aa9041aed346369e10316 100644
--- a/src/headerssync.cpp
+++ b/src/headerssync.cpp
@@ -21,9 +21,14 @@ constexpr size_t HEADER_COMMITMENT_PERIOD{610};
 // 14521/610 = ~23.8 commitments
 constexpr size_t REDOWNLOAD_BUFFER_SIZE{14521};
 
-// Our memory analysis assumes 48 bytes for a CompressedHeader (so we should
-// re-calculate parameters if we compress further)
-static_assert(sizeof(CompressedHeader) == 48);
+// On Bitcoin, this would be 48 bytes. There, the memory analysis assumes 48
+// bytes for a CompressedHeader (so we would have to re-calculate parameters if
+// we were to compress further).
+//
+// On Dogecoin, the compression is mostly void due to the auxpow. We have to add
+// it anyway here, which results in 16 extra bytes for the std::shared_ptr.
+// In the future, we can find a more efficient way of storing the auxpow.
+static_assert(sizeof(CompressedHeader) == 64);
 
 HeadersSyncState::HeadersSyncState(NodeId id,
                                    const Consensus::Params &consensus_params,
diff --git a/src/headerssync.h b/src/headerssync.h
index 2f23dee5ab802cb3d6ee65ab64331adac22e2531..64a9247d3a61843c6f3d97048670645c76103b7a 100644
--- a/src/headerssync.h
+++ b/src/headerssync.h
@@ -25,6 +25,11 @@ struct CompressedHeader {
     uint32_t nTime{0};
     uint32_t nBits{0};
     uint32_t nNonce{0};
+    // AuxPow of the header.
+    // Note: this mostly voids the advantage of the compressed header (since
+    // AuxPow can be 100s of bytes), but there's no simple way of avoiding this
+    // loss in efficiency without a lot of re-engineering.
+    std::shared_ptr<CAuxPow> auxpow;
 
     CompressedHeader() { hashMerkleRoot.SetNull(); }
 
@@ -34,6 +39,7 @@ struct CompressedHeader {
         nTime = header.nTime;
         nBits = header.nBits;
         nNonce = header.nNonce;
+        auxpow = header.auxpow;
     }
 
     CBlockHeader GetFullHeader(const BlockHash &hash_prev_block) {
@@ -44,6 +50,7 @@ struct CompressedHeader {
         ret.nTime = nTime;
         ret.nBits = nBits;
         ret.nNonce = nNonce;
+        ret.auxpow = auxpow;
         return ret;
     };
 };
diff --git a/src/test/dogecoin_auxpow_methods_tests.cpp b/src/test/dogecoin_auxpow_methods_tests.cpp
index cbaaf0109f7cafcd04694bd2f20783d8e20e2b92..52e372d00837f1e5e6fee63595e5a5221063b85d 100644
--- a/src/test/dogecoin_auxpow_methods_tests.cpp
+++ b/src/test/dogecoin_auxpow_methods_tests.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <headerssync.h>
 #include <pow/auxpow.h>
 #include <pow/pow.h>
 #include <primitives/auxpow.h>
@@ -54,12 +55,22 @@ BOOST_AUTO_TEST_CASE(auxpow_blockheader_setnull_test) {
 
     BOOST_CHECK(VersionHasAuxPow(header.nVersion));
     BOOST_CHECK(header.auxpow);
+    {
+        CompressedHeader compressedHeader(header);
+        BOOST_CHECK(compressedHeader.auxpow);
+        BOOST_CHECK(compressedHeader.GetFullHeader(BlockHash()).auxpow);
+    }
 
     // Test SetNull also resets the auxpow
     header.SetNull();
 
     BOOST_CHECK(!VersionHasAuxPow(header.nVersion));
     BOOST_CHECK(!header.auxpow);
+    {
+        CompressedHeader compressedHeader(header);
+        BOOST_CHECK(!compressedHeader.auxpow);
+        BOOST_CHECK(!compressedHeader.GetFullHeader(BlockHash()).auxpow);
+    }
 }
 
 BOOST_AUTO_TEST_SUITE_END()
